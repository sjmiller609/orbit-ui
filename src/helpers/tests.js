import { BrowserRouter } from 'react-router-dom'
import { shape } from 'prop-types'
import { mount } from 'enzyme'

// Instantiate router context
const router = {
  history: new BrowserRouter().history,
  route: {
    location: {},
    match: {},
  },
}

const createContext = () => ({
  context: { router },
  childContextTypes: { router: shape({}) },
})

// wrap with router context - to mount RR links correctly
export function mountWrap(node) {
  return mount(node, createContext())
}
