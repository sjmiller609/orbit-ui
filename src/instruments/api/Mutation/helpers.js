const has = (str, find) => ~str.toLowerCase().indexOf(find.toLowerCase())

export const formErrors = ({ error, errors }) => {
  if (!error) return

  let err = JSON.stringify(error).toLowerCase()
  return errors.find(e => {
    if (e.key && has(err, e.key)) return e
    else if (e.allKeys && e.allKeys.every(k => has(err, k))) return e
    else if (e.someKeys && e.someKeys.some(k => has(err, k))) return e
  })
}
