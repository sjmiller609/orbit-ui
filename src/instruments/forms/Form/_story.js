import React from 'react'
import { storiesOf } from '@storybook/react'
import { Form } from 'instruments'

storiesOf('Instruments|Forms.Form', module).add('Default', () => (
  <Form onSubmit={() => null}>Form Children</Form>
))
