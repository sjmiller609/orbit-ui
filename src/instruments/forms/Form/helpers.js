// faltten and stringify nested keys
export const unpack = (obj, ns, hash) => {
  if (!obj) return
  return Object.entries(obj).reduce((acc, [key, value]) => {
    const keypath = ns ? `${ns}.${key}` : key
    if (value && typeof value === 'object') {
      return unpack(value, keypath, acc)
    }
    acc[keypath] = value
    return acc
  }, hash || {})
}

const roll = (keys, obj, value) => {
  const key = keys[0]
  // case 1 - not obj
  if (keys.length === 1) {
    obj[key] = value
    return obj
  }
  // case 2 - doesn't exist
  if (!obj[key]) {
    obj[key] = isNaN(keys[1]) ? {} : []
  }
  // case 3 - exists
  obj[key] = roll(
    keys.slice(1),
    Array.isArray(obj[key]) ? [...obj[key]] : { ...obj[key] },
    value
  )

  return obj
}

// convert back into obj
export const pack = (obj, type = {}) => {
  return Object.entries(obj).reduce((acc, [key, value]) => {
    const keys = key.split('.')

    return roll(keys, acc, value)
  }, type)
}

export const packChild = ({ name, obj }) => {
  const obj2 = {}
  let isArray
  Object.entries(obj).forEach(k => {
    if (name.length < k[0].length && k[0].indexOf(name) === 0) {
      const n = k[0].slice(name.length + 1)
      // must be true for all matching
      isArray = !isNaN(n.slice(0, n.indexOf('.')))
      obj2[n] = k[1]
    }
  })
  if (Object.keys(obj2).length) return pack(obj2, isArray ? [] : {})
  return null
}

export const removeChild = ({ name, obj = {} }) => {
  const obj2 = {}
  Object.entries(obj).forEach(k => {
    if (name.length >= k[0].length || k[0].indexOf(name) !== 0) {
      obj2[k[0]] = k[1]
    }
  })
  return obj2
}
