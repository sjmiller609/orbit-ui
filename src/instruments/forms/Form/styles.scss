@import '../../../styles/mixins.scss';

.form {
}
