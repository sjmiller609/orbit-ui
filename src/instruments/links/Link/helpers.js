// check if url is external
export const externalUrl = url => {
  if (typeof url === 'object') return false
  const path = /^https?:\/\/|^\/\//i
  if (path.test(url) && !~url.indexOf(window.location.origin)) return true
  return
}
