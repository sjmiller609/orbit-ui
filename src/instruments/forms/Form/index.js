'use strict'
import React from 'react'
import PropTypes from 'prop-types'

import { jsonEqual } from 'helpers/compare'
import { unpack, pack } from './helpers'
import UnsavedChangesAlert from '../UnsavedChangesAlert'
import s from './styles.scss'

const errorField = name => 'error_' + name

const Form = FormComponent => {
  class Form extends React.Component {
    onSubmit = this.onSubmit.bind(this)
    update = this.update.bind(this)
    updateErrors = this.updateErrors.bind(this)
    checkSave = this.checkSave.bind(this)
    checkErrors = this.checkErrors.bind(this)
    field = this.field.bind(this)
    unpack = this.unpack.bind(this)
    data = this.unpack(this.props.data)

    state = {
      data: this.data,
      save: false,
      submitted: false,
    }

    componentWillReceiveProps({ data, error }) {
      if (!jsonEqual(data, this.props.data)) {
        this.data = this.unpack(this.props.data)
        this.setState({ data: this.data })
      }
      if (!jsonEqual(error, this.props.error) && error)
        this.updateErrors(error.name, error.error)
    }

    shouldComponentUpdate(nextProps, nextState) {
      if (!jsonEqual(this.state, nextState)) return true
      if (!jsonEqual(this.props, nextProps)) return true
      return false
    }

    // turn all objects into keys with string names with dot notation
    unpack(data) {
      if (!data || typeof data !== 'object') return data
      return unpack(data)
    }

    getValue(name) {
      const { data } = this.state
      if (!~name.indexOf('.')) return data[name]
      const n = name.split('.')
      if (data[n[0]]) return data[n[0]][n[1]]
      else return null
    }

    update(key, value) {
      const set = {
        data: {
          ...this.state.data,
          [key]: value,
        },
        submitted: false,
      }
      this.setState(set)
    }

    // updates all validation checks
    updateErrors(name, error) {
      // NOTE: have to namespace this way because state is updated simulatenously by all fields, so they get overwriten
      const set = {}
      if (error !== this.state[errorField(name)]) set[errorField(name)] = error
      // check save after validation check
      const save = this.checkSave({ ...this.state, ...set })
      if (typeof save === 'boolean') set.save = save
      this.setState(set)
    }

    checkSave(state) {
      const { saveOnLoad } = this.props
      const save = state.save
      // changes made
      const equal = saveOnLoad ? false : jsonEqual(state.data, this.data)

      // has errors
      if (!this.checkErrors(state)) return false

      if (save && equal) return false
      if (!save && !equal) return true
      return
    }

    // return true if all required fields have data and are valid
    // and if isn't required but has data and is valid
    checkErrors(state) {
      const keys = Object.keys(state)
      return !keys.some(field => {
        if (field.indexOf(errorField('')) !== 0) return
        if (state[field]) return true
      })
    }

    onSubmit(e) {
      e.preventDefault()
      const { saveOnLoad, onSubmit } = this.props
      const { save, data } = this.state
      if (!save) return
      if (!saveOnLoad) this.setState({ save: false, submitted: true })
      onSubmit(pack(data), this.updateErrors)
    }

    field(name) {
      // If a field is named with dot notation, will convert into object
      return {
        name,
        value: this.state.data[name],
        error: this.state[errorField(name)],
        updateErrors: this.updateErrors,
        onChange: this.update,
        submitted: this.state.submitted,
      }
    }

    render() {
      const { save } = this.state
      /* eslint-disable no-unused-vars  */
      // don't pass down some props
      const { data, onSubmit, ...otherProps } = this.props

      const newProps = {
        ...otherProps,
        form: {
          save,
          // return an object with props for that field
          field: this.field,
        },
      }

      return (
        <form onSubmit={this.onSubmit} className={s.form}>
          <FormComponent {...newProps} />
          <UnsavedChangesAlert alert={save} />
        </form>
      )
    }
  }

  Form.propTypes = {
    error: PropTypes.object,
    data: PropTypes.object,
    onSubmit: PropTypes.func,
    children: PropTypes.element,
    saveOnLoad: PropTypes.bool, // if true, will enable saving preloaded data
  }

  Form.defaultProps = {
    data: {},
  }

  return Form
}

export default Form
