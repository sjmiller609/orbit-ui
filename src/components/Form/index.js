import React, {Component} from 'react'

import styles from './styles.module.css'

class Form extends Component {
  formSubmit = e => {
    const { handleSubmit } = this.props;
    e.preventDefault();
    handleSubmit(e);
  }

  submitErrors = (error) => {
    if(error != null && error != {}) {
      const errors = JSON.parse(error).graphQLErrors;
      if (errors.length > 0) {
        return errors.map((e, i) =>
          <p className={styles.error} key={`error-${i}`}>⚠️ {e.message}</p>
        )
      }
    }
  }

  render() {
    const { children, handleSubmit, error } = this.props;

    return (
      <form onSubmit={e => this.formSubmit(e)}>
        {children}
      </form>
    )
  }
}

export default Form
