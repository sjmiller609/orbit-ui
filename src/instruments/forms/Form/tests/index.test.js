import React from 'react'
import { mountWrap } from 'helpers/tests'
import Form from '../index'

const props = {
  error: {},
  data: {},
  onSubmit: jest.fn(),
  children: React.Fragment,
  saveOnLoad: true,
  alert: true,
}

describe('Form', () => {
  it('renders correctly', () => {
    const wrapper = mountWrap(<Form {...props} />)
    expect(wrapper).toMatchSnapshot()
  })
})
