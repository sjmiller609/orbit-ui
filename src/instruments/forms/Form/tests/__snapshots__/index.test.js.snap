// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form renders correctly 1`] = `
<Form
  alert={true}
  data={Object {}}
  error={Object {}}
  onSubmit={[Function]}
  saveOnLoad={true}
/>
`;
